R/row_count.R

```diff
@@ -0,0 +1,101 @@
+#' possible to make "type safe" comparisons using the `exact` argument, where
+#' `"2" == 2` is not treated as identical.
+#'
+#' @param data A data frame with at least two columns, where number of specific
+#' values are counted row-wise.
+#' @param count The value for which the row sum should be computed. May be a
+#' numeric value, a character string (for factors or character vectors), `NA` or
+#' `Inf`.
+#' @param exact Logical, if `TRUE`, `count` matches only values of same type
+#' (i.e. when `count = 2`, the value `"2"` is not counted and vice versa).
+#' By default, when `exact = FALSE`, `count = 2` also matches `"2"`. See
+#' 'Examples'.
+#'
+#' @inheritParams extract_column_names
+#' @inheritParams row_means
+#'
+#' @return A vector with row-wise counts of values specified in `count`.
+#'
+#' @examples
+#' dat <- data.frame(
+#'   c1 = c(1, 2, NA, 4),
+#'   c2 = c(NA, 2, NA, 5),
+#'   c3 = c(NA, 4, NA, NA),
+#'   c4 = c(2, 3, 7, 8)
+#' )
+#'
+#' # count all 2s per row
+#' row_count(dat, count = 2)
+#' # count all missing values per row
+#' row_count(dat, count = NA)
+#'
+#' dat <- data.frame(
+#'   c1 = c("1", "2", NA, "3"),
+#'   c2 = c(NA, "2", NA, "3"),
+#'   c3 = c(NA, 4, NA, NA),
+#'   c4 = c(2, 3, 7, Inf)
+#' )
+#' # count all 2s and "2"s per row
+#' row_count(dat, count = 2)
+#' # only count 2s, but not "2"s
+#' row_count(dat, count = 2, exact = TRUE)
+#'
+#' @export
+row_count <- function(data,
+                      select = NULL,
+                      exclude = NULL,
+                      count = NULL,
+                      exact = FALSE,
+                      ignore_case = FALSE,
+                      regex = FALSE,
+                      verbose = TRUE) {
+  # evaluate arguments
+  select <- .select_nse(select,
+    data,
+    exclude,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+
+  if (is.null(count)) {
+    insight::format_error("`count` must be a valid value (including `NA` or `Inf`), but not `NULL`.")
+  }
+
+  if (is.null(select) || length(select) == 0) {
+    insight::format_error("No columns selected.")
+  }
+
+  data <- .coerce_to_dataframe(data[select])
+
+  # check if we have a data framme with at least two columns
+  if (ncol(data) < 2) {
+    insight::format_error("`data` must be a data frame with at least two numeric columns.")
+  }
+
+  # special case: count missing
+  if (is.na(count)) {
+    rowSums(is.na(data))
+  } else {
+    # comparisons in R using == coerce values into a atomic vector, i.e.
+    # 2 == "2" is TRUE. If `exact = TRUE`, we only want 2 == 2 or "2" == "2".
+    # to achieve this, we simply compute the comparison on numeric or non-numeric
+    # columns only
+    if (isTRUE(exact)) {
+      numeric_columns <- vapply(data, is.numeric, TRUE)
+      if (is.numeric(count)) {
+        data <- data[numeric_columns]
+      } else {
+        data <- data[!numeric_columns]
+      }
+    }
```

Comment:
This doesn't take into account other coercions, such as:

```r
as.Date("2020-01-01") == "2020-01-01"
#> TRUE
```
Maybe the value of `count` gives the type to look for and then we only retain columns that match this type?

Reply:
What would you suggest how to solve this?

```r
typeof(as.Date("2020-01-01"))
#> [1] "double"
```

The type of Date variables is double, but it compares to character vector (like in your example). Using the `class()` instead is difficult for factor-comparison, because `count = "a"` should also match factor levels `a` (else, it's hard to define that value), so we can't compare classes, too.

comparisons of types is difficult, comparisons of class is difficult.

Reply:
But if we're in the strict case, should we count both a character "a" and a factor level "a"?

Reply:
How would you count factor levels then?

Reply:
(if you only want to count specific factor level, how would you define the `count` argument?)

Reply:
Maybe an extra arg `use_levels`? I don't have a perfect solution for this for now, but my point is that if we have  `allow_coercion = FALSE` then it should prevent all coercion, not only numeric <-> non-numeric.

Reply:
An extra argument is probably too much. But I just thought that `count = factor("a")` may work?

Reply:
Yes, sounds good to me